From: Daniel Kahn Gillmor <dkg@fifthhorseman.net>
Date: Sun, 24 Jun 2018 19:52:26 -0400
Subject: Clean up Autocrypt code when promises fail.

Enigmail's code invokes reject() but then doesn't call return.  It's
possible that this causes some other code to execute or trigger side
effects, so it's generally better to explicitly return afterward.
---
 package/autocrypt.jsm | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/package/autocrypt.jsm b/package/autocrypt.jsm
index 36283bc..45302a7 100644
--- a/package/autocrypt.jsm
+++ b/package/autocrypt.jsm
@@ -71,6 +71,7 @@ var EnigmailAutocrypt = {
       }
       catch (ex) {
         reject("processAutocryptHeader error " + ex);
+        return;
       }
       let foundTypes = {};
       let paramArr = [];
@@ -301,6 +302,7 @@ var EnigmailAutocrypt = {
         function onError(error) {
           EnigmailLog.DEBUG("autocrypt.jsm: getOpenPGPKeyForEmail: could not open database\n");
           reject("getOpenPGPKeyForEmail1 error " + error);
+          return;
         }
       ).then(
         function _f() {
@@ -312,7 +314,7 @@ var EnigmailAutocrypt = {
           conn.close();
 
           if (resultObj.data.length === 0) {
-            resolve(null);
+            return resolve(null);
           }
           else {
             let retArr = [];
@@ -327,13 +329,14 @@ var EnigmailAutocrypt = {
               });
             }
 
-            resolve(retArr);
+            return resolve(retArr);
           }
         }
       ).
       catch((err) => {
         conn.close();
         reject("getOpenPGPKeyForEmail2 error " + err);
+        return;
       });
     });
   },
@@ -419,6 +422,7 @@ var EnigmailAutocrypt = {
       }).catch(e => {
         EnigmailLog.DEBUG("autocrypt.jsm: createSetupMessage: error " + e + "\n");
         reject(2);
+        return;
       });
     });
   },
@@ -455,6 +459,7 @@ var EnigmailAutocrypt = {
         }
         else {
           reject(99);
+          return;
         }
       });
     });
@@ -501,6 +506,7 @@ var EnigmailAutocrypt = {
         }
         else {
           reject("getSetupMessageData");
+          return;
         }
       });
 
@@ -551,15 +557,18 @@ var EnigmailAutocrypt = {
             }
             else {
               reject("keyImportFailed");
+              return;
             }
           });
         }
         else {
           reject("keyImportFailed");
+          return;
         }
       }).
       catch(err => {
         reject("wrongPasswd");
+        return;
       });
     });
   },
